pylabrobot/pumps/calibration.py

```diff
@@ -0,0 +1,181 @@
+
+    Returns:
+      PumpCalibration
+
+    Raises:
+      NotImplementedError: if the calibration filetype or format is not supported.
+      ValueError: if num_items is not specified when calibration is a value.
+    """
+
+    if calibration is None:
+      return PumpCalibration.uncalibrated()
+    if isinstance(calibration, dict):
+      return PumpCalibration.load_from_dict(calibration)
+    if isinstance(calibration, list):
+      return PumpCalibration.load_from_list(calibration)
+    if isinstance(calibration, (float, int)):
+      if num_items is None:
+        raise ValueError("num_items must be specified if calibration is a value.")
+      return PumpCalibration.load_from_value(calibration, num_items)
+    if isinstance(calibration, str):
+      if calibration.endswith(".json"):
+        return PumpCalibration.load_from_json(calibration)
+      if calibration.endswith(".csv"):
+        return PumpCalibration.load_from_csv(calibration)
+      raise NotImplementedError("Calibration filetype not supported.")
+    return PumpCalibration()
+
+  @classmethod
+  def load_from_json(cls, file: str) -> PumpCalibration:
+    """ Load a calibration from a json file.
+
+    Args:
+      file: json file to load calibration from.
+
+    Returns:
+      PumpCalibration
+
+    Raises:
+      TypeError: if the calibration pulled from the json is not a dictionary or list.
+    """
+    with open(file, "rb") as f:
+      calibration = json.load(f)
+    if isinstance(calibration, dict):
+      calibration = {int(key): float(value) for key, value in calibration.items()}
+      return PumpCalibration.load_from_dict(calibration)
+    if isinstance(calibration, list):
+      return PumpCalibration(calibration)
+    raise TypeError(f"Calibration pulled from {file} is not a dictionary or list.")
+
+  @classmethod
+  def load_from_csv(cls, file: str) -> PumpCalibration:
+    """ Load a calibration from a csv file.
+
+    Args:
+      file: csv file to load calibration from.
+
+    Returns:
+      PumpCalibration
+    """
+
+    with open(file, encoding="utf-8", newline="") as f:
+      reader = csv.DictReader(f)
+      calibration = {int(row["0"]): float(row["1"]) for row in reader}
+    return PumpCalibration.load_from_dict(calibration)
+
+  @classmethod
+  def load_from_dict(cls, calibration: Dict[int, Union[int, float]]) -> PumpCalibration:
+    """ Load a calibration from a dictionary.
+
+    Args:
+      calibration: dictionary to load calibration from.
+
+    Returns:
+      PumpCalibration
+
+    Raises:
+      ValueError: if the calibration dictionary is not formatted correctly.
+    """
+
+    if not any(key == 0 for key in calibration.keys()):
+      if any(key == 1 for key in calibration.keys()):
+        calibration = {key - 1: value for key, value in calibration.items()}
+      else:
+        raise ValueError("Calibration dictionary keys must start at 0 or 1.")
+    if len(calibration) == 1:
+      calibration = {0: list(calibration.values())[0]}
```

Comment:
Given the code above, this wouldn't change the list. When the user inputs a singular format (always starting with a 0, or a 1) this should no longer be needed.